cmd/cerbosctl/hub/store/upload_git.go

```diff
@@ -0,0 +1,314 @@
+						AddOrUpdate: &storev1.File{
+							Path:     change.name,
+							Contents: contents,
+						},
+					},
+				})
+			case OpDelete:
+				batch = append(batch, &storev1.FileOp{
+					Op: &storev1.FileOp_Delete{
+						Delete: change.name,
+					},
+				})
+			default:
+				yield(nil, fmt.Errorf("unexpected operation %s", change.operation))
+			}
+
+			batchCounter++
+			if batchCounter == modifyFilesBatchSize {
+				if !yield(batch, nil) {
+					return
+				}
+
+				batch = make([]*storev1.FileOp, 0, modifyFilesBatchSize)
+				batchCounter = 0
+			}
+		}
+
+		if batchCounter > 0 {
+			yield(batch, nil)
+		}
+	}
+}
+
+func (ugc *UploadGitCmd) diff(r *git.Repository, from, to plumbing.Hash) (*diff, error) {
+	fromCommit, err := r.CommitObject(from)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get commit object for commit %q: %w", from.String(), err)
+	}
+
+	toCommit, err := r.CommitObject(to)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get commit object for commit %q: %w", to.String(), err)
+	}
+
+	isAncestor, err := fromCommit.IsAncestor(toCommit)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if ancestor: %w", err)
+	}
+
+	if !isAncestor {
+		return nil, fmt.Errorf("commit %q is not ancestor of commit %q", to.String(), from.String())
+	}
+
+	fromTree, err := fromCommit.Tree()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tree of commit %q: %w", from.String(), err)
+	}
+
+	toTree, err := toCommit.Tree()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tree of commit %q: %w", to.String(), err)
+	}
+
+	objectChanges, err := object.DiffTree(fromTree, toTree)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get diff tree: %w", err)
+	}
+
+	changes, err := ugc.changes(objectChanges)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get changes: %w", err)
+	}
+
+	return &diff{hash: to, changes: changes}, nil
+}
+
+func (ugc *UploadGitCmd) changes(objectChanges object.Changes) ([]*change, error) {
+	pathToRepo, err := filepath.Abs(ugc.Path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get absolute path to repository: %w", err)
+	}
+
+	changes := make([]*change, 0, len(objectChanges))
+	for _, objectChange := range objectChanges {
+		from, to, err := objectChange.Files()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get file from change: %w", err)
+		}
+
+		var nameOfToBeDeleted string
+		var name string
+		var operation op
+		switch {
+		case from != nil && to != nil:
+			if objectChange.From.Name == objectChange.To.Name {
+				name = objectChange.To.Name
+				operation = OpAddOrUpdate
+			} else {
+				name = objectChange.To.Name
+				operation = OpAddOrUpdate
+				nameOfToBeDeleted = objectChange.From.Name
+			}
+		case from == nil && to != nil:
+			name = objectChange.To.Name
+			operation = OpAddOrUpdate
+		case from != nil:
+			name = objectChange.From.Name
+			operation = OpDelete
+		}
+
+		if ugc.skipped(name) {
+			continue
+		}
```

Comment:
Move this after line 256 because it's safer to do the subdirectory check after the path has been cleaned.